### PR TITLE
WIP: Fix ChatPacketHandler setting wrong Ids to Database.

### LINKED ISSUE #249: BUG: ChatServer tries to set wrong Player Ids into `friends`-Table
### Make sure you've done the following:

- [X] I have checked that this issue has not already been reported.

- [X] I have validated that this issue is not a syntax error of either MySQL or SQLite.

- [X] I have pulled the latest version of the main branch of DarkflameServer and have confirmed that the issue exists there.

### Issue Description

When I try to add a Friend it throws a Foreign Key Error: 
```
terminating with uncaught exception of type sql::SQLException: Cannot add or update a child row: a foreign key constraint fails (`darkflame`.`friends`, CONSTRAINT `friends_ibfk_1` FOREIGN KEY (`player_id`) REFERENCES `charinfo` (`id`))
```

When I disabled the Foreign Key Checks with `SET FOREIGN_KEY_CHECKS=0` before the Server runs they Query, this is what it writes to the Database: 

Screenshot 2021-12-16 at 22 41 05

## To get the Foreign Key Error:

- Run the latest Version of the Server
- Join the Server with a Friend or a second Client
- Ask this User to be your Friend
- The Log will output the Error written in the Description

## To get the Data to be written to the DB:

- In dChatServer/ChatPacketHandler.cpp in Line 134 add:

```cpp
auto stmtDFKC = Database::CreatePreppedStmt("SET FOREIGN_KEY_CHECKS=0");
stmtDFKC->execute();
delete stmtDFKC;
```

- Build the Server and start it
- Join the Server with a Friend or a second Client
- Ask this User to be your Friend
- The Friend-System works until one changes the World

### Expected Behavior

The Server should write the right User Ids to the Database so that the Friend and Team System will work.

### Environment

MacBook Pro 2019 (Intel Core i9, 16GB Ram, macOS 12.1)
Apple clang version 13.0.0 (clang-1300.0.29.30)

### PR DESCRIPTION
**Pull request by l-mbert**
Fixes #249

When the playerId get's casted to a uInt64 it changes to an objectId (?) for Example `123456789`, when set with only `setInt` it sets the right userId into the friends Table and the Foreign Key Constraint is fullfilled.

Also the Reason why the ChatServer is crashing when someone sends a Friend Request.

Tested with a Friend on a local Server on MacOS 12.

### CODE REVIEW

**Comment by Jettford** [comment]
Player IDs are LWOOBJIDs (Just a type def for a uint64), they are supposed to be 64 bitt integers and should be stored as such?

**Comment by l-mbert** [comment]
Then the Foreign Key Constraint in the Database is wrong. 
Since it references the id in the `charinfo` Table and will always fail if those aren't matching.

**Comment by leolion3** [comment]
@Jettford In the queries used by the Chat and Friends systems the values are used unchecked and interchangeably, which is why the Chat server keeps crashing. At some points in the database queries you are using setInt(playerID) on the statements being executed, in others you are using setUInt(playerID), all the while the logger is formatting the player IDs as unsigned longs in its print statements which is why the bug was being missed.

**Comment by Jettford** [comment]
> Then the Foreign Key Constraint in the Database is wrong.
Since it references the id in the charinfo Table and will always fail if those aren't matching.
In dGame/UserManager.cpp Line 325 objectID is set as a UInt, so was the creation of the charinfos' already wrong?

Ah, they must store them without the LWOOBJID flags set, my bad

**Comment by l-mbert** [comment]
> Ah, they must store them without the LWOOBJID flags set, my bad

@Jettford The question is: is this the right fix? Smarter would be to change `dGame/UserManager.cpp` to `setUInt64` but this would break all charinfo Ids now since they are Ints.

What we could do is when we get the charinfo, we could add a check if the id casted to an UInt64 matches our expected Id and update the charinfo accordingly.

**Comment by DarwinAnim8or** [comment]
IIRC DLU always just stored player LWOOBJID's straight to the SQL Db as they'd get chopped up and only the real ID (not the half we use for flags) would remain.

The thing is, when I made the table for friends, it didn't have any constraint at all, but before we went public that got changed and now we're having crashes.

**Comment by l-mbert** [comment]
What would you like me to do here? I'm struggling a bit to find the right Solution.

My fix currently also has another Error where the Friends aren't updating their Online- and World change status, which I need to look into.

We could either fix the Ids in the `charinfo`-Tables since they are set wrong, we could use this and I'd need to fix the OnlineState thing or we could remove the constraint.

**Comment by jgkawell** [comment]
> What we could do is when we get the charinfo, we could add a check if the id casted to an UInt64 matches our expected Id and update the charinfo accordingly.

Yeah this one is tricky since data is not out of whack. I like the above idea of putting check in the code to cleanup the data for anyone upgrading from old versions.

**Comment by Xiphoseer** [comment]
IMO this is the correct fix and I'd argue that a sql migration is probably the easiest way to fix this for existing instances.

To elaborate on
> as they'd get chopped up and only the real ID (not the half we use for flags) would remain.

The ObjectIDs have two 32 bit parts. The first one is a bitset that describes what kind of ID it is and should always be the same for characters. So moving friends to match chardata is probably a good idea.

As for the fix, something like
```sql
UPDATE friends SET player_id = player_id % 0x100000000, friend_id = friend_id % 0x100000000
```

might work well

File: dChatServer/ChatPacketHandler.cpp [hunk 1]
```diff
@@ -23,17 +23,18 @@ void ChatPacketHandler::HandleFriendlistRequest(Packet* packet) {
 
 	//Get our friends list from the Db:
 	auto stmt = Database::CreatePreppedStmt("SELECT * FROM friends WHERE player_id = ? OR friend_id = ?");
-	stmt->setUInt64(1, playerID);
-	stmt->setUInt64(2, playerID);
+	stmt->setInt(1, playerID);
+	stmt->setInt(2, playerID);
```
[the remark this answers is not in the document]
Comment (l-mbert) [reply dChatServer/ChatPacketHandler.cpp:27]: We could apply this. The Friend System still seems to be broken, it doesn't register any changes in Worlds or Online for the User and the Friend.

**Comment by l-mbert** [comment]
I tried fixing `HandleFriendlistRequest` with getting an UInt64 in line 35 and 37 but of course the ID would remain the same and not convert into an LWOOBJID.

The problem we now have with this fix is, that the `SendFriendUpdate` will always fail because we supply it with the wrong ID / aren't able to get the player in line 56 and if the player changes worlds or goes offline/online it will not be registered.

Also, somehow the Player which didn't initiate the friend request will get themselves as a Friend if the Server restarts.

Maybe someone has an idea how me might fix this?

**Comment by AxtonF** [comment]
Has there been any update to this? And there is one more thing I was wondering, maybe someone could help me here. I tried setting adding friends manually in the database by adding char1_id chard2_id and char2_id chard1_id manually to the friends table, but they still didn't show up in the ingame friendlist. Is there a workaround to add friends manually?

**Comment by l-mbert** [comment]
Hey @Xiphoseer I'm currently working on my Finals and haven't had time to look into it more and will be gone for about 2-3 months.

What would be the best going forward? Close it?

Maybe somebody else could adopt it if they want.

**Comment by Xiphoseer** [comment]
> Hey @Xiphoseer I'm currently working on my Finals and haven't had time to look into it more and will be gone for about 2-3 months.
> 
> What would be the best going forward? Close it?
> 
> Maybe somebody else could adopt it if they want. 

All good. You can mark it as WIP or closed whatever you feel like. It's a small change so it's unlikely you (or someone else) needs to completely re-do it next time around so we can probably reuse the PR then

**Comment by EmosewaMC** [comment]
@DarwinAnim8or Could you shed some light on a direction forward here?  Should the friends table be shifted to contain charIDs or should the friends table continue to contain LWOOBJIDs?  I recall the latter being how you mentioned this is done and lots more would need to change if we made the shift from LWOOBJIDs to charIDs since iirc the chatserver currently just gets the LWOOBJID from the friends table and now a conversion would need to happen.

**Comment by EmosewaMC** [comment]
Closing to address the issue properly in a different PR.

**Pull request closed without merging**